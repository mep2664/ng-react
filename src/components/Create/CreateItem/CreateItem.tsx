import * as React from "react";
import { NotFound } from "../..";
import { CreateProject } from "../CreateProject";
import { CreateSprint } from "../CreateSprint";
import { CreateTeam } from "../CreateTeam";
import { CreateTicket } from "../CreateTicket"

interface ICreateItem {
    item: string;
}

export const CreateItem: React.FC<ICreateItem> = ({ item }) => {
    switch (item) {
        case "project":
            return <CreateProject />;
        case "sprint":
            return <CreateSprint />;
        case "team":
            return <CreateTeam />;
        case "ticket":
            return <CreateTicket />;
        default:
            return <NotFound />;
    }
}