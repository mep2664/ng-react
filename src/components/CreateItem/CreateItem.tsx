import * as React from "react";
import { SelectInput, TextInput } from "../../components";
import styled from "styled-components";
import gql from "graphql-tag";
import { useMutation } from "@apollo/react-hooks";

const CREATE_TICKET = gql`
    mutation createTicket($projectName: String!, $description: String, $priority: String,
        $sprintId: Int, $storyPoints: Int, $ticketType: String) {
            createTicket(projectName: $projectName, description: $description, priority: $priority,
                sprintId: $sprintId, storyPoints: $storyPoints, ticketType: $ticketType)
            {
                ticket {
                    ticketId
                    projectName
                    ticketNumber
                    description
                    priority
                    sprintId
                    storyPoints
                    ticketType
                }
            }
    }
`;

const TicketInfo = styled.div`
    display: grid;
    grid-gap: 25px;
    align-items: center;
    justify-items: end;
    grid-template-columns: minmax(auto, 300px) auto;
    box-sizing: border-box;
    padding: 25px;
`;

const SelectInputs = styled.div`
    display: grid;
    grid-gap: 10px;
    align-items: center;
    justify-items: end;
    grid-template-columns: auto minmax(auto, 300px);
`;

const DescriptionWrapper = styled.div`
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 100%;
`;

const Description = styled.textarea`
    width: 100%;
    height: 100%;
    box-sizing: border-box;
    padding: 5px;
    resize: none;
`;

const projectOptions = [
    { caption: "Polaris", value: "POLARIS" },
    { caption: "Pitcher", value: "PITCHER" },
];

const sprintOptions = [
    { caption: "Sprint-One", value: 1 },
    { caption: "Sprint-Two", value: 2 },
];

const typeOptions = [
    { caption: "Enhancement", value: "Enhancement" },
    { caption: "Bug", value: "Bug" },
    { caption: "Research", value: "Research" },
];

const priorityOptions = [
    { caption: "Blocker", value: "Blocker" },
    { caption: "Standard", value: "Standard" },
    { caption: "Minor", value: "Minor" },
];

const storyPointsOptions = () => {
    const sequence: number[] = [1, 2];
    for (let idx = 0; idx < 5; idx++) {
        sequence.push(sequence[idx] + sequence[idx + 1]);
    }
    const options: any[] = [];
    for (let numb in sequence) {
        options.push({ caption: sequence[numb].toString(), value: sequence[numb] });
    }
    return options;
}

export const CreateItem: React.FC = () => {
    const [projectName, setProject] = React.useState<string>("");
    const [sprintId, setSprintId] = React.useState<string>("");
    const [ticketType, setTicketType] = React.useState<string>("");
    const [priority, setPriority] = React.useState<string>("");
    const [storyPoints, setStoryPoints] = React.useState<number>(0);
    const [description, setDescription] = React.useState<string>("");
    const [createTicket] = useMutation(CREATE_TICKET);

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        const data = {
            projectName,
            sprintId,
            ticketType,
            priority,
            storyPoints,
            description,
        }
        console.log(data);
        createTicket({ variables: data });

    }

    return (
        <div>
            {`Create Ticket...`}
            <TicketInfo>
                <form action="localhost:5556/ticket" method="post" onSubmit={handleSubmit}>
                    <SelectInputs>
                        <SelectInput label="Project" name="project" options={projectOptions} value={projectName} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setProject(e.target.value)} />
                        <SelectInput label="Sprint" name="sprint" options={sprintOptions} value={sprintId} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setSprintId(e.target.value)} />
                        <SelectInput label="Type" name="type" options={typeOptions} value={ticketType} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setTicketType(e.target.value)} />
                        <SelectInput label="Priority" name="priority" options={priorityOptions} value={priority} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setPriority(e.target.value)} />
                        <SelectInput label="Story Points" name="storyPoints" options={storyPointsOptions()} value={storyPoints} onChange={(e: React.ChangeEvent<HTMLSelectElement>) => setStoryPoints(Number(e.target.value))} />
                    </SelectInputs>
                    <DescriptionWrapper>
                        <label>Description</label>
                        <Description placeholder="Enter a ticket description here..." value={description} onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => setDescription(e.target.value)} />
                    </DescriptionWrapper>
                    <input type="submit" value="submit" />
                </form>
            </TicketInfo>
        </div>
    );
}