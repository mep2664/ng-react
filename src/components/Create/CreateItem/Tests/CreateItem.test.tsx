import React from "react"
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect";
import { CreateItem } from "../CreateItem"

jest.mock("../../CreateProject", () => {
    return {
        CreateProject: () => <span>CREATE_PROJECT</span>,
    }
});

jest.mock("../../CreateSprint", () => {
    return {
        CreateSprint: () => <span>CREATE_PROJECT</span>,
    }
});

jest.mock("../../CreateTeam", () => {
    return {
        CreateTeam: () => <span>CREATE_PROJECT</span>,
    }
});

jest.mock("../../CreateTicket", () => {
    return {
        CreateTicket: () => <span>CREATE_PROJECT</span>,
    }
});

jest.mock("../../../NotFound", () => {
    return {
        NotFound: () => <span>NOT_FOUND</span>,
    }
});

describe("Tests for the CreateItem component", () => {
    test("Component matches snapshot", () => {
        const { container } = render(
            <CreateItem
                item="not-found"
            />,
        );
        expect(container).toMatchSnapshot();
    });

    test("Project case returns something", () => {
        const { queryByText } = render(
            <CreateItem
                item="project"
            />,
        );
        expect(queryByText("CREATE_PROJECT")).toBeDefined();
    });

    test("Sprint case returns something", () => {
        const { queryByText } = render(
            <CreateItem
                item="sprint"
            />,
        );
        expect(queryByText("CREATE_SPRINT")).toBeDefined();
    });

    test("Team case returns something", () => {
        const { queryByText } = render(
            <CreateItem
                item="team"
            />,
        );
        expect(queryByText("CREATE_TEAM")).toBeDefined();
    });

    test("Ticket case returns something", () => {
        const { queryByText } = render(
            <CreateItem
                item="ticket"
            />,
        );
        expect(queryByText("CREATE_TICKET")).toBeDefined();
    });
});
