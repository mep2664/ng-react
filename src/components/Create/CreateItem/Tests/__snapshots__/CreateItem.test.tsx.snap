// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for the CreateItem component Component matches snapshot 1`] = `
<div>
  <span>
    NOT_FOUND
  </span>
</div>
`;
